File: examples/python/03_edit/main.py
"""03_edit — Edit a document by replacing placeholder text.

Creates a DOCX from Markdown that contains {{NAME}} and {{DATE}},
opens it with EditableDocument, replaces both placeholders, saves to
a second temp file, then reads back and verifies. Exit 0 on success.
"""
import sys
import tempfile
import os

from office_oxide import Document, EditableDocument, create_from_markdown


MARKDOWN = """\
# Invoice

Dear {{NAME}},

Please find attached your invoice for services rendered on {{DATE}}.

## Summary

- Service: Office Oxide Pro License
- Amount: $499.00
- Due date: 30 days from {{DATE}}

Thank you for your business!
"""


def main() -> None:
    template_tmp = None
    output_tmp = None
    try:
        with tempfile.NamedTemporaryFile(suffix=".docx", delete=False) as f:
            template_tmp = f.name
        with tempfile.NamedTemporaryFile(suffix=".docx", delete=False) as f:
            output_tmp = f.name

        # Create template
        create_from_markdown(MARKDOWN, "docx", template_tmp)

        # Edit
        with EditableDocument.open(template_tmp) as ed:
            n1 = ed.replace_text("{{NAME}}", "Alice Smith")
            n2 = ed.replace_text("{{DATE}}", "2026-04-26")
            ed.save(output_tmp)

        print(f"Replacements: {{{{NAME}}}} x{n1}, {{{{DATE}}}} x{n2}")

        # Verify
        with Document.open(output_tmp) as doc:
            text = doc.plain_text()

        assert "Alice Smith" in text, "name replacement failed"
        assert "2026-04-26" in text, "date replacement failed"
        assert "{{NAME}}" not in text, "placeholder still present"
        assert "{{DATE}}" not in text, "placeholder still present"

        print("Edit verified successfully.")
        print("--- final text ---")
        print(text)
    finally:
        for p in [template_tmp, output_tmp]:
            if p and os.path.exists(p):
                os.unlink(p)


if __name__ == "__main__":
    try:
        main()
    except Exception as exc:
        print(f"error: {exc}", file=sys.stderr)
        sys.exit(1)

File: examples/go/03_edit/main.go
//go:build office_oxide_dev

// 03_edit — Edit a document by replacing placeholder text.
//
// Creates a DOCX from Markdown with {{NAME}} and {{DATE}} placeholders,
// opens with oo.OpenEditable, replaces both, saves to a second temp file,
// then reads back and verifies. Exit 0 on success.
//
// Build and run (inside the monorepo, after `cargo build --release --lib`):
//
//	go run -tags office_oxide_dev ./examples/go/03_edit/
package main

import (
	"fmt"
	"log"
	"os"
	"strings"

	oo "github.com/yfedoseev/office_oxide/go"
)

const markdown = `# Invoice

Dear {{NAME}},

Please find attached your invoice for services rendered on {{DATE}}.

## Summary

- Service: Office Oxide Pro License
- Amount: $499.00
- Due date: 30 days from {{DATE}}

Thank you for your business!
`

func main() {
	// Create temp files
	tpl, err := os.CreateTemp("", "oo_03_template_*.docx")
	if err != nil {
		log.Fatal(err)
	}
	tplPath := tpl.Name()
	tpl.Close()
	defer os.Remove(tplPath)

	out, err := os.CreateTemp("", "oo_03_output_*.docx")
	if err != nil {
		log.Fatal(err)
	}
	outPath := out.Name()
	out.Close()
	defer os.Remove(outPath)

	// Create template
	if err := oo.CreateFromMarkdown(markdown, "docx", tplPath); err != nil {
		log.Fatalf("CreateFromMarkdown: %v", err)
	}

	// Edit
	ed, err := oo.OpenEditable(tplPath)
	if err != nil {
		log.Fatalf("OpenEditable: %v", err)
	}
	n1, err := ed.ReplaceText("{{NAME}}", "Alice Smith")
	if err != nil {
		log.Fatalf("ReplaceText NAME: %v", err)
	}
	n2, err := ed.ReplaceText("{{DATE}}", "2026-04-26")
	if err != nil {
		log.Fatalf("ReplaceText DATE: %v", err)
	}
	if err := ed.Save(outPath); err != nil {
		log.Fatalf("Save: %v", err)
	}
	ed.Close()

	fmt.Printf("Replacements: {{NAME}} x%d, {{DATE}} x%d\n", n1, n2)

	// Verify
	doc, err := oo.Open(outPath)
	if err != nil {
		log.Fatalf("Open output: %v", err)
	}
	defer doc.Close()

	text, err := doc.PlainText()
	if err != nil {
		log.Fatalf("PlainText: %v", err)
	}

	if !strings.Contains(text, "Alice Smith") {
		log.Fatal("name replacement failed: 'Alice Smith' not found")
	}
	if !strings.Contains(text, "2026-04-26") {
		log.Fatal("date replacement failed: '2026-04-26' not found")
	}
	if strings.Contains(text, "{{NAME}}") {
		log.Fatal("placeholder {{NAME}} still present")
	}
	if strings.Contains(text, "{{DATE}}") {
		log.Fatal("placeholder {{DATE}} still present")
	}

	fmt.Println("Edit verified successfully.")
	fmt.Println("--- final text ---")
	fmt.Println(text)
}
